Add a confmap.Provider that communicates with the EEC (#175)
Dynatrace/dynatrace-otel-collector: confmap/provider/eecprovider/README.md

# EEC Provider

This is an OpenTelemetry Collector `confmap.Provider` module that allows the
Collector to be configured with the Dynatrace Extensions Execution Controller.

> [!WARNING]

> > This is not intended for direct customer use, but is only intended for use by

Reviewer: ```suggestion
> This is an internal component not intended for direct customer use, but is only intended for use by
```

> > This is not intended for customer use, but is only intended for use by
> > Dynatrace when managing its own Collector deployments. As such, issues with
> > this component will fall outside our support policy.

Reviewer 2: I think "only intended for use by Dynatrace when managing its own Collector deployments" makes it sound like it's only for use within Dynatrace the SaaS cluster or something like that. It's not clear that it's for use by Dynatrace in customer environments.

```suggestion
> This is not intended for direct customer use, but is only intended for use by
> Collector installations installed and managed by Dynatrace. Configuring this
> component directly is not supported.
```

Author: Yeah I wasn't really satisfied with my wording. I like yours better, thanks.

## Configuration

This confmap Provider can be configured by pasing query parameter-formatted values
inside the fragment of the URL given to the config flag. For example:

> ```text

Reviewer 3: ```suggestion
```shell
```

Author: I tried this, but the problem I ran into is that the config fragments are highlighted as comments, at least in my local Markdown renderer. I'll give it a shot with the GitHub renderer and see if it's any different.

Author: You can see what it looks like here: https://github.com/Dynatrace/dynatrace-otel-collector/commit/eec2b66eb3dd4d3c8bc3b62853a88decebed54ad?short_path=5b27679#diff-5b27679011dbb54249c6f87e02849959e30682d84058764b1a8bedfd3957c6c7

I think the highlighted ampersand is more distracting than anything, I'd like to opt to leave it as plain text for now. Let me know if you think the highlighting would still be worth it.

Reviewer 3: it's just a nitpick. I leave the decision on you.

dynatrace-otel-collector --config=eec://my.eec.host:31098#refresh-interval=5s&auth-header=Api-Token&auth-file=/var/private/token.key
```

### Options

| Key | Default | Description |
|-----|---------|-------------|

> | auth-env |  None | An environment variable that should be read to get a plaintext API token or other key to be used in an HTTP header. Mutually exclusive from `auth-file`, passing both options should result in an error. |

Reviewer 4: "should result in an error" or will result in an error? 🤔

Author: Good catch. "Will result in an error" is right.

> | auth-file | None | A filepath containing a plaintext version of an API token or other key to be used in an HTTP header to authenticate with the EEC. |

Reviewer 3: `auth-header` is present in the example, but missing here, should we add it?

Author: I actually meant to get rid of `auth-header`. I'll remove it from the example.

| auth-header | Api-Token | A header that will be used to authenticate the provider with the EEC. |
| refresh-interval | 10s | A time duration that defines how frequently the provider should check the given URL for updates. |

> | insecure | false | If set to "true", use HTTP for the connection to the server. If unset or set to "false", use HTTPS. |

Reviewer 2: I think we should leave use out of public documentation if we're going to discourage direct use? We can either document internally or rely on the code comments in the config struct

Author: I'm fine with whatever, this just seemed like the ideal place to have config and the code comments may not be easy to locate. I think any sufficiently-motivated user could figure it out, so I was just going to rely on our disclaimer to discourage users. That said, I can strip down this file if we think it's a risk.
